File: front/web/lib/red-wave.ts
/**
 * The red-wave (vague de rouge) impact projection — the Workbench /red-wave source (AIDOS step S22).
 *
 * KRD §42 (la vague de rouge IS EXACTLY the set of stale links after a bump — computed, never
 * hunted; it STARTS at the mirror then cascades to the projections api/db/types/operation/action/
 * button) · §98 (the ChangeSet example: a bump fires PostKernelChange: fire-red-wave --from-mirror →
 * mirrors red first, projections after) · §74 (the hooks.yaml line PostKernelChange: rehash &&
 * fire-red-wave --from-mirror) · §49.4 (RedWorkQueue/RedWorkItem: target, reason ∈ {version_stale,
 * failed_test, incident}, status open|claimed|blocked|resolved) · §112 (load-bearing vs cosmetic —
 * a cosmetic change does not propagate; ADR 0020) · ADR 0020 (this step's decisions).
 *
 * This module is the DECLARED projection of the Go package back/runtime/redwave — the same closure,
 * the same mirror-first ordering, the same load-bearing propagation rule, reusing the same staleness
 * check as back/kernel/links.Resolve — so the /red-wave panel orders + colours each wave item exactly
 * as the Go Impact / Enqueue compute them. One source, no drift.
 *
 * DETERMINISM-FIRST (CLAUDE.md §6/§8): pure functions over their input — no clock, no rng, no I/O —
 * so the same (bumped, edges, heads) always yields the same ordered wave. The reproducibility mirror
 * lib/red-wave.test.ts (fast-check) pins determinism, wave==set-of-stale-links, mirror-first order,
 * no-propagation-on-cosmetic, the |wave| enqueue count, and totality.
 *
 * READ-ONLY (CLAUDE.md §7 ui-completeness, the wall): /red-wave PROJECTS the queue/wave; it never
 * writes truth and never enqueues. Enqueuing is the harness-invoked PostKernelChange hook; applying a
 * change goes via propose → ChangeSet → approval, never from this screen.
 */

/** A PINNED layer reference: an id plus the concrete version it points at (id@version). */
export interface Ref {
	id: string;
	version: string;
}

/** The six KRD §41 link kinds (the closed set, mirrors back/kernel/links.Kind). */
export type LinkKind =
	| "projects_to"
	| "derives_from"
	| "contracts_with"
	| "triggers"
	| "binds"
	| "mirrors";

/** An S17 versioned link (kind + pinned from/to), mirrors links.Link. */
export interface Link {
	kind: LinkKind;
	from: Ref;
	to: Ref;
}

/** The staleness verdict of a link against heads (mirrors links.LinkStatus). */
export type LinkStatus = "green" | "stale" | "absent";

/** heads maps a target id to its current head version (mirrors links.Heads). */
export type Heads = Record<string, string>;

/** The render-grouping layer of a target (mirrors redwave.Layer). */
export type Layer = "mirror" | "projection" | "operation_action" | "button";

/** Why a RedWorkItem is red (the closed §49.4 set, mirrors redwave.Reason). */
export type Reason = "version_stale" | "failed_test" | "incident";

/** The lifecycle status of a RedWorkItem (the closed §49.4 set, mirrors redwave.Status). */
export type Status = "open" | "claimed" | "blocked" | "resolved";

/**
 * A propagating edge of the wave's input graph: an S17 link PLUS the DECLARED load-bearing flag
 * (the composes weight, S18/S19 §112; ADR 0020) and a render Layer hint. Mirrors redwave.Edge.
 */
export interface Edge {
	link: Link;
	loadBearing: boolean;
	layer: Layer;
}

/** One element of the wave (KRD §49.4), mirrors redwave.RedWorkItem. */
export interface RedWorkItem {
	target: string;
	reason: Reason;
	dependencies: string[];
	layer: Layer;
	waveId?: string;
}

/** The ordered set of RedWorkItems a bump triggers (KRD §42), mirrors redwave.RedWave. */
export interface RedWave {
	items: RedWorkItem[];
}

/**
 * resolve is the PURE staleness check (mirrors links.Resolve): absent when heads has no entry for
 * the target; green when pinned exactly to the head; stale otherwise. Total, deterministic, never
 * throws. Keys only on the pinned `to` ref.
 */
export function resolve(link: Link, heads: Heads): LinkStatus {
	const head = heads[link.to.id];
	if (head === undefined) return "absent";
	return head === link.to.version ? "green" : "stale";
}

/** The canonical mirror-first rank: mirror(0) < projection(1) < operation_action(2) < button(3). */
function layerRank(l: Layer): number {
	switch (l) {
		case "mirror":
			return 0;
		case "projection":
			return 1;
		case "operation_action":
			return 2;
		case "button":
			return 3;
		default:
			return 4;
	}
}

/**
 * impact computes the red wave a bump triggers (KRD §42; mirrors redwave.Impact). It is a total,
 * deterministic function of (bumped, edges, heads):
 *   - the wave is the transitive closure of edges whose target is already red AND resolve reports
 *     stale|absent AND the edge is LOAD-BEARING (a cosmetic edge never propagates, §112);
 *   - it walks OUTWARD from the bumped sources (mirror first);
 *   - it is ORDERED mirror-first, then by target id (byte-stable — determinism).
 * Pure: no clock, no rng, no I/O. Never throws.
 */
export function impact(
	bumped: readonly string[],
	edges: readonly Edge[],
	heads: Heads,
): RedWave {
	if (bumped.length === 0) return { items: [] };

	const red = new Set<string>(bumped);
	const items = new Map<
		string,
		{ target: string; reason: Reason; layer: Layer; deps: Set<string> }
	>();

	for (;;) {
		let grew = false;
		for (const e of edges) {
			if (resolve(e.link, heads) === "green") continue;
			if (!red.has(e.link.to.id)) continue;
			if (!e.loadBearing) continue;
			const from = e.link.from.id;
			let it = items.get(from);
			if (it === undefined) {
				it = {
					target: from,
					reason: "version_stale",
					layer: e.layer,
					deps: new Set(),
				};
				items.set(from, it);
			}
			it.deps.add(e.link.to.id);
			if (!red.has(from)) {
				red.add(from);
				grew = true;
			}
		}
		if (!grew) break;
	}

	const out: RedWorkItem[] = [...items.values()].map((it) => ({
		target: it.target,
		reason: it.reason,
		dependencies: [...it.deps].sort(),
		layer: it.layer,
	}));
	out.sort((a, b) => {
		const ra = layerRank(a.layer);
		const rb = layerRank(b.layer);
		if (ra !== rb) return ra - rb;
		return a.target < b.target ? -1 : a.target > b.target ? 1 : 0;
	});
	return { items: out };
}

/**
 * enqueue stamps every wave item with the waveId (the bump hash) and returns the rows VALUE — it
 * writes nothing (the wall). Returns exactly |wave| rows. Mirrors redwave.Enqueue.
 */
export function enqueue(w: RedWave, waveId: string): RedWorkItem[] {
	return w.items.map((it) => ({ ...it, waveId }));
}

/** isEmpty reports whether the wave carries no items (no bump ⇒ empty wave, §42). */
export function isEmpty(w: RedWave): boolean {
	return w.items.length === 0;
}

File: back/runtime/redwave/redwave.go
// Package redwave is the pure RED-WAVE (vague de rouge) impact engine of the AIDOS Runtime
// (KRD §42, §74, §98). After a kernel hash bump, it computes the set of stale links and
// failing mirrors the bump triggers — the worklist — and drains it into a RedWorkQueue (§49.4).
//
// THE LAW (KRD §42): the red wave IS EXACTLY the set of stale links after a bump — nothing
// more, nothing less. It is COMPUTED, never hunted. It STARTS at the mirror (the bumped
// source's `mirrors` reflection reddens first) and CASCADES to the projections that
// derive from / project to the changed source (api/db/types/operation/action/button), in
// that order (mirror-first, §42/§98). Changing an entity reddens api/db/types; changing a
// button reddens its view IFF the button is LOAD-BEARING (a cosmetic change does not redden
// the view — KRD §112).
//
// REUSE, DON'T REINVENT (CLAUDE.md §6; ADR 0020): the staleness of each edge is decided by
// S17's links.Resolve — this package NEVER re-implements or forks it. An item appears in the
// wave iff Resolve reports stale|absent for its link; there is no item without a stale link
// and no stale link without an item (§42). "Load-bearing" is the DECLARED composes weight
// (S18/S19, §112), carried on the edge as a bool — not an S17 link field, not invented here.
//
// PURE (CLAUDE.md §6 determinism-first): Impact and Enqueue are total, deterministic functions
// of their input — no DB, no clock, no rng, no I/O. The bumped set, the link graph and the
// heads are READ from the arguments handed in, never fetched, so the same input yields a
// byte-identical ordered wave and the wave is replayable. The rapid property mirror pins
// determinism, wave==set-of-stale-links, mirror-first order, no-propagation-on-cosmetic, the
// |wave| enqueue count, and totality (no panic on a malformed/absent target).
//
// READ-ONLY against truth (the wall, CLAUDE.md §2): this package writes nothing. Enqueue
// returns the rows VALUE; the actual INSERT into runtime.red_work_queue (below the waterline)
// is done by the harness-invoked PostKernelChange hook through the agent's INSERT+SELECT
// grant — never above the line.
package redwave

import (
	"sort"

	"github.com/steph-frtech/aidos/back/kernel/links"
)

// Reason is why a RedWorkItem is red — the CLOSED set of KRD §49.4. version_stale is the
// red-wave seed (a link pinned to a non-head / absent version); failed_test and incident are
// the other on-ramps (a mirror that went red on a run, a production incident). S22 computes
// version_stale; the other two are referenced (their producers are later steps).
type Reason string

const (
	// ReasonVersionStale — the item's link is pinned to a non-head/absent version (§42 seed).
	ReasonVersionStale Reason = "version_stale"
	// ReasonFailedTest — the item's mirror went red on a run (a later producer).
	ReasonFailedTest Reason = "failed_test"
	// ReasonIncident — the item was opened by a production incident (a later producer).
	ReasonIncident Reason = "incident"
)

// Status is a RedWorkItem's lifecycle state — the CLOSED set of KRD §49.4. S22 only ever
// produces `open`; the transitions (claimed/blocked/resolved) are the SCHEDULER's job (a
// later step, §49.4), not computed here.
type Status string

const (
	// StatusOpen — the item is in the queue, unclaimed (the only status S22 produces).
	StatusOpen Status = "open"
	// StatusClaimed — a scheduler leased the item to an agent (a later step).
	StatusClaimed Status = "claimed"
	// StatusBlocked — the item is blocked on a dependency (a later step).
	StatusBlocked Status = "blocked"
	// StatusResolved — the item's projection was turned green (a later step).
	StatusResolved Status = "resolved"
)

// Layer is the coarse render layer of a target — used by the Workbench to GROUP the wave
// (mirror → api/db/types → operation/action → button). It is a render hint, not part of the
// closure logic; the order of the wave is decided by the closure, not by the layer.
type Layer string

const (
	// LayerMirror — the bumped source's mirror reflection (reddens FIRST, §42/§98).
	LayerMirror Layer = "mirror"
	// LayerProjection — an api/db/types projection (derives_from/projects_to the source).
	LayerProjection Layer = "projection"
	// LayerOperationAction — an operation or action layer.
	LayerOperationAction Layer = "operation_action"
	// LayerButton — a control/button or its view.
	LayerButton Layer = "button"
)

// Edge is one propagating edge of the wave's input graph: an S17 link PLUS the DECLARED
// load-bearing flag (the composes weight, S18/S19 §112; ADR 0020) and a render Layer hint.
// The link's staleness is decided ONLY by links.Resolve (reused, never forked). LoadBearing
// gates propagation: a cosmetic edge (LoadBearing=false) never propagates red beyond itself.
// A nil/empty Layer is treated as LayerProjection by classification, never a crash.
type Edge struct {
	// Link is the S17 versioned link (kind + pinned from/to). Resolve(Link, heads) decides red.
	Link links.Link `json:"link"`
	// LoadBearing is the DECLARED composes weight (true = load-bearing/critical, false =
	// cosmetic), S18/S19 §112. A cosmetic edge does NOT propagate red to its consumer.
	LoadBearing bool `json:"load_bearing"`
	// Layer is the render-grouping hint for the consuming `from` side (mirror/projection/…).
	Layer Layer `json:"layer"`
}

// RedWorkItem is one element of the wave (KRD §49.4): a target (a mirror_id or a projection
// ref), why it is red, its declared dependencies, plus the render Layer and the bump
// (wave_id) that opened it. Target is the consuming `from` side — the layer that must be
// turned green. It is the queue row's value before it is enqueued.
type RedWorkItem struct {
	// Target is the consuming layer ref that is now red (a mirror_id or a projection id@version).
	Target string `json:"target"`
	// Reason is why the item is red (version_stale for the wave seed).
	Reason Reason `json:"reason"`
	// Dependencies are the targets this item depends on (the upstream red targets, ordered).
	Dependencies []string `json:"dependencies"`
	// Layer is the render-grouping hint.
	Layer Layer `json:"layer"`
	// WaveID is the bump's content hash — the id of the wave this item belongs to (set at Enqueue).
	WaveID string `json:"wave_id"`
}

// RedWave is the ORDERED set of RedWorkItems a bump triggers (KRD §42). The order is the
// contract: the bumped source's mirror(s) come FIRST, then the projections (mirror-first,
// §42/§98). An empty wave (no bump, or no stale link) is the zero case.
type RedWave struct {
	// Items are the wave's RedWorkItems in canonical (mirror-first) order.
	Items []RedWorkItem `json:"items"`
}

// IsEmpty reports whether the wave carries no items (no bump ⇒ empty wave, §42).
func (w RedWave) IsEmpty() bool { return len(w.Items) == 0 }

// layerRank gives the canonical mirror-first ordering rank of a layer: mirror(0) <
// projection(1) < operation_action(2) < button(3). An unknown layer ranks last (after the
// known layers) so a malformed layer never jumps ahead of the mirror.
func layerRank(l Layer) int {
	switch l {
	case LayerMirror:
		return 0
	case LayerProjection:
		return 1
	case LayerOperationAction:
		return 2
	case LayerButton:
		return 3
	default:
		return 4
	}
}

// Impact computes the red wave a bump triggers (KRD §42, the PURE engine). It is a total,
// deterministic function of (bumped, edges, heads):
//
//   - bumped : the set of kernel source ids the bump changed (the new head moved). Empty ⇒
//     empty wave (no bump ⇒ nothing to do, §42).
//   - edges  : the versioned link graph (S17 links + the declared load-bearing flag + layer).
//   - heads  : the current head version per target id (the S17 Heads input; the bumped
//     sources' heads have already moved here, so their consumers' pinned versions are stale).
//
// The wave is the transitive closure of edges whose target resolves stale|absent, walked
// OUTWARD from the mirror. Each edge's staleness is decided by links.Resolve (REUSED, never
// forked). A LOAD-BEARING edge propagates red to its consumer; a COSMETIC edge does not (its
// downstream beyond itself stays green, §112). The result is ORDERED mirror-first.
//
// PURE: no DB, no clock, no rng. NEVER PANICS on a malformed/absent target (Resolve is total;
// a malformed layer ranks last). Same input ⇒ byte-identical ordered wave (the property pins it).
func Impact(bumped []string, edges []Edge, heads links.Heads) RedWave {
	if len(bumped) == 0 {
		return RedWave{}
	}

	// The set of currently-red TARGET ids (the upstream layers a bump made stale). We seed it
	// with the bumped sources themselves, then grow it by the closure: an edge becomes red iff
	// its `to` target is already red AND links.Resolve reports it stale|absent AND the edge is
	// load-bearing (cosmetic edges do not propagate). The consuming `from` side then becomes a
	// red target in turn, so a chain (entity → mirror → projection) cascades.
	red := make(map[string]bool, len(bumped))
	for _, id := range bumped {
		red[id] = true
	}

	// Collect the wave items. We iterate the closure to a fixpoint (the graph is finite and
	// edges only ADD red targets, so it terminates). itemFor keys an item by its target so a
	// target reddened by two paths yields ONE item with merged, ordered dependencies.
	type pending struct {
		target string
		reason Reason
		layer  Layer
		deps   map[string]bool
	}
	items := map[string]*pending{}

	for {
		grew := false
		for _, e := range edges {
			// Reuse S17 Resolve — the single source of staleness truth.
			status := links.Resolve(e.Link, heads)
			if status == links.StatusGreen {
				continue // the link is pinned to the live head — not red.
			}
			// The wave is EXACTLY the set of stale links, but a link only PROPAGATES from an
			// already-red target. The link's `to` (target) must be red for its `from` (consumer)
			// to redden — this is the outward-from-the-source walk.
			if !red[e.Link.To.ID] {
				continue
			}
			// Cosmetic edges do not propagate red to their consumer (§112; ADR 0020).
			if !e.LoadBearing {
				continue
			}
			from := e.Link.From.ID
			it, seen := items[from]
			if !seen {
				it = &pending{target: from, reason: ReasonVersionStale, layer: e.Layer, deps: map[string]bool{}}
				items[from] = it
			}
			it.deps[e.Link.To.ID] = true
			if !red[from] {
				red[from] = true
				grew = true
			}
		}
		if !grew {
			break
		}
	}

	// Materialize the items into the ordered wave. Order: mirror-first by layer rank, then by
	// target id for a stable, byte-identical result (determinism — no map-iteration leak).
	out := make([]RedWorkItem, 0, len(items))
	for _, it := range items {
		deps := make([]string, 0, len(it.deps))
		for d := range it.deps {
			deps = append(deps, d)
		}
		sort.Strings(deps)
		out = append(out, RedWorkItem{
			Target:       it.target,
			Reason:       it.reason,
			Dependencies: deps,
			Layer:        it.layer,
		})
	}
	sort.SliceStable(out, func(i, j int) bool {
		ri, rj := layerRank(out[i].Layer), layerRank(out[j].Layer)
		if ri != rj {
			return ri < rj
		}
		return out[i].Target < out[j].Target
	})
	return RedWave{Items: out}
}

// Enqueue materializes a wave into RedWorkQueue rows (KRD §49.4). It is a PURE projection: it
// stamps every item with the waveID (the bump's content hash) and returns the rows VALUE —
// it writes NOTHING (the wall; the harness-invoked hook does the INSERT below the line). The
// returned items all carry the wave's id; the caller persists them with status=open,
// owner_agent=NULL, lease_until=NULL. It returns exactly |wave| rows.
func Enqueue(w RedWave, waveID string) []RedWorkItem {
	rows := make([]RedWorkItem, len(w.Items))
	for i, it := range w.Items {
		it.WaveID = waveID
		rows[i] = it
	}
	return rows
}
